client/alloc_watcher.go

```diff
@@ -0,0 +1,542 @@
+				return ctx.Err()
+			}
+		}
+		if resp.Alloc == nil {
+			p.logger.Printf("[DEBUG] client: blocking alloc %q has been GC'd", p.prevAllocID)
+			return nil
+		}
+		if resp.Alloc.Terminated() {
+			// Terminated!
+			p.nodeID = resp.Alloc.NodeID
+			return nil
+		}
+
+		// Update the query index and requery.
+		if resp.Index > req.MinQueryIndex {
+			req.MinQueryIndex = resp.Index
+		}
+	}
+
+	if done() {
+		return ctx.Err()
+	}
+
+	return nil
+}
+
+// Migrate alloc data from a remote node if the new alloc has migration enabled
+// and the old alloc hasn't been GC'd.
+func (p *remotePrevAlloc) Migrate(ctx context.Context, dest *allocdir.AllocDir) error {
+	if !p.migrate {
+		// Volume wasn't configured to be migrated, return early
+		return nil
+	}
+
+	p.waitingLock.Lock()
+	p.migrating = true
+	p.waitingLock.Unlock()
+	defer func() {
+		p.waitingLock.Lock()
+		p.migrating = false
+		p.waitingLock.Unlock()
+	}()
+
+	p.logger.Printf("[DEBUG] client: alloc %q copying from remote previous alloc %q", p.allocID, p.prevAllocID)
+
+	if p.nodeID == "" {
+		// NodeID couldn't be found; likely alloc was GC'd
+		p.logger.Printf("[WARN] client: alloc %q couldn't migrate data from previous alloc %q; previous alloc may have been GC'd",
+			p.allocID, p.prevAllocID)
+		return nil
+	}
+
+	addr, err := p.getNodeAddr(ctx, p.nodeID)
+	if err != nil {
+		return err
+	}
+
+	prevAllocDir, err := p.migrateAllocDir(ctx, addr)
+	if err != nil {
+		return err
+	}
+
+	if err := dest.Move(prevAllocDir, p.tasks); err != nil {
+		// cleanup on error
+		prevAllocDir.Destroy()
+		return err
+	}
+
+	if err := prevAllocDir.Destroy(); err != nil {
+		p.logger.Printf("[ERR] client: error destroying allocdir %q: %v", prevAllocDir.AllocDir, err)
+	}
+	return nil
+}
+
+// getNodeAddr gets the node from the server with the given Node ID
+func (p *remotePrevAlloc) getNodeAddr(ctx context.Context, nodeID string) (string, error) {
+	req := structs.NodeSpecificRequest{
+		NodeID: nodeID,
+		QueryOptions: structs.QueryOptions{
+			Region:     p.config.Region,
+			AllowStale: true,
+		},
+	}
+
+	resp := structs.SingleNodeResponse{}
+	for {
+		err := p.rpc.RPC("Node.GetNode", &req, &resp)
+		if err != nil {
+			p.logger.Printf("[ERR] client: failed to query node info %q: %v", nodeID, err)
+			retry := getAllocRetryIntv + lib.RandomStagger(getAllocRetryIntv)
+			select {
+			case <-time.After(retry):
+				continue
+			case <-ctx.Done():
+				return "", ctx.Err()
+			}
+		}
+		break
+	}
+
+	if resp.Node == nil {
+		return "", fmt.Errorf("node %q not found", nodeID)
+	}
+
+	scheme := "http://"
+	if resp.Node.TLSEnabled {
+		scheme = "https://"
+	}
+	return scheme + resp.Node.HTTPAddr, nil
+}
+
+// migrate a remote alloc dir to local node
+func (p *remotePrevAlloc) migrateAllocDir(ctx context.Context, nodeAddr string) (*allocdir.AllocDir, error) {
+	// Create the previous alloc dir
+	prevAllocDir := allocdir.NewAllocDir(p.logger, filepath.Join(p.config.AllocDir, p.prevAllocID))
```

Comment:
This method does on errors otherwise the caller does. Clarified in the docstring.